client-v2/src/shared/util/shared.ts

```diff
@@ -1,17 +1,111 @@
-import { CollectionWithNestedArticles } from 'shared/types/Collection';
+import { CollectionWithNestedArticles, Group } from 'shared/types/Collection';
 import { selectors as collectionSelectors } from 'shared/bundles/collectionsBundle';
 import { selectSharedState } from 'shared/selectors/shared';
 import { State } from 'types/State';
 
 import { normalize, denormalize } from './schema';
+import { CollectionConfig } from 'types/FaciaApi';
+import v4 from 'uuid/v4';
+import keyBy from 'lodash/keyBy';
+
+const createGroup = (
+  id: string,
+  name: string | null,
+  articleFragments: string[] = []
+): Group => ({
+  id,
+  name,
+  uuid: v4(),
+  articleFragments
+});
+
+const getUUID = <T extends { uuid: string }>({ uuid }: T) => uuid;
+
+// this is horrible but is the only way to do it!
+const groupByIndex = (groups: Group[], index: number): Group | undefined =>
+  groups.find(g => parseInt(g.id || '0', 10) === index);
+
+const getAllArticleFragments = (groups: Group[]) =>
+  groups.reduce(
+    (acc, { articleFragments: afs }) => [...acc, ...afs],
+    [] as string[]
+  );
+
+const addMissingGroupsForStage = (
+  groupIds: string[],
+  entities: { [id: string]: Group },
+  collectionConfig: CollectionConfig
+) => {
+  const groups = groupIds.map(id => entities[id]);
+  const newGroups = collectionConfig.groups
+    ? collectionConfig.groups.map((name, i) => {
+        const existingGroup = groupByIndex(groups, i);
+
+        return existingGroup
+          ? { ...existingGroup, name }
+          : createGroup(`${i}`, name);
+      })
+    : [createGroup('0', null, getAllArticleFragments(groups))];
+
+  return {
+    newGroups: keyBy(newGroups, getUUID),
+    groupIds: newGroups
+      .map(getUUID)
+      .slice()
+      .reverse()
+  };
+};
+
+interface ReduceResult {
+  live: string[];
+  draft: string[];
+  previously: string[];
+  newGroups: { [key: string]: Group };
+}
+
+const addMissingGroups = (
+  normalisedCollection: any,
+  collectionConfig: CollectionConfig
+) =>
+  (['live', 'previously', 'draft'] as ['live', 'previously', 'draft']).reduce(
+    (acc, key) => {
+      const { newGroups, groupIds } = addMissingGroupsForStage(
+        normalisedCollection.result[key],
+        normalisedCollection.entities.groups,
+        collectionConfig
+      );
+      return {
+        ...acc,
+        newGroups: {
+          ...acc.newGroups,
+          ...newGroups
+        },
+        [key]: groupIds
+      };
+    },
+    { live: [], draft: [], previously: [], newGroups: {} } as ReduceResult
+  );
 
 const normaliseCollectionWithNestedArticles = (
-  collection: CollectionWithNestedArticles
+  collection: CollectionWithNestedArticles,
+  collectionConfig: CollectionConfig
 ) => {
   const normalisedCollection = normalize(collection);
+  const { newGroups, live, draft, previously } = addMissingGroups(
```

Comment:
They're missing in a `Collection` because they only exist in the `CollectionConfig` so basically this merges the groups from a collection config into a collection so that we don't have to use the collection config anywhere else. Does that make sense?

Reply:
@Reettaphant I've actually changed the name of the function and a return key so that it hopefully make it clearer 😃

Reply:
Yes, that makes sense now. This looks fine, you might want to call it sth like `addGroupsFromConfig` but it's not a big deal.